File: python3/367.valid-perfect-square.py
#
# @lc app=leetcode id=367 lang=python3
#
# [367] Valid Perfect Square
#

# @lc code=start
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        lo, hi = 0, num
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if mid*mid < num: lo = mid+1
            else: hi = mid
        return int(lo)*int(lo) == num
        
# @lc code=end

File: java/367.valid-perfect-square.java
/*
 * @lc app=leetcode id=367 lang=java
 *
 * [367] Valid Perfect Square
 */

// @lc code=start
class Solution {
    public boolean isPerfectSquare(int num) {
        int l = 1, r = num;
        while (l < r) {
            int mid = l + (r - l) / 2;
            long square = (long) mid * mid;
            if (square < num) l = mid + 1;
            else r = mid;
        }

        return l * l == num;
    }
}
// @lc code=end

